File: cubic-chart/src/main/java/com/lyhv/cubic_chart/HistogramSeekbarView.java
package com.lyhv.cubic_chart;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.TimeInterpolator;
import android.animation.ValueAnimator;
import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Path;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.content.res.AppCompatResources;

import com.github.mikephil.charting.animation.ChartAnimator;
import com.github.mikephil.charting.animation.Easing;
import com.github.mikephil.charting.charts.LineChart;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;
import com.github.mikephil.charting.interfaces.dataprovider.LineDataProvider;
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet;
import com.github.mikephil.charting.renderer.LineChartRenderer;
import com.github.mikephil.charting.utils.Transformer;
import com.github.mikephil.charting.utils.ViewPortHandler;
import com.google.android.material.animation.AnimationUtils;

import java.text.NumberFormat;
import java.util.List;

public class HistogramSeekbarView extends LinearLayout {
    private static final TimeInterpolator ANIMATION_INTERPOLATOR = Easing.EaseInOutSine;
    private static final NumberFormat INTEGER_FORMAT = NumberFormat.getIntegerInstance();
    private final ValueAnimator histogramTransitionAnimator = ValueAnimator.ofFloat(0.0f, 1.0f);
    public TextView debugInfo;
    public LineChart histogram;
    public double lastRealMax = 1.0d;
    public double lastRealMin = 0.0d;
    /* access modifiers changed from: private */
    public LineDataSet previousDataSet;
    //  public Rang<Double> rangeSeekbar;
    private Drawable activeGraphFillDrawable;
    // private IValueFormatter debugGraphFormatter = $$Lambda$HistogramSeekbarView$jMW_yJHc2oIkaWKYTjeleCHrp3c.INSTANCE;
    private Drawable graphGradientDrawable;
    private GraphInactiveFillRenderer graphInactiveFillRenderer;
    private int graphValueTextSize;
    private int inactiveGraphColor;
    private double lastSeekbarMaxValue = 1.0d;
    private double lastSeekbarMinValue = 0.0d;
    //  private RangeSeekBar noListingsDisabledDummySeekbar;
    private int viewPortLeftOffset;
    private int viewPortRightOffset;

    public HistogramSeekbarView(Context context) {
        super(context);
        initializeHistogram();
    }

    public HistogramSeekbarView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        initializeHistogram();
    }

    private void initializeHistogram() {
        this.histogram.setData(new LineData());
        this.histogram.setTouchEnabled(false);
        this.histogram.setDescription(null);
        this.histogram.setDrawGridBackground(false);
        this.histogram.setDrawBorders(false);
        this.histogram.setDrawMarkers(false);
        this.histogram.getLegend().setEnabled(false);
        this.histogram.setNoDataText(null);
        this.histogram.setHardwareAccelerationEnabled(true);
        this.histogram.getAxisLeft().setDrawGridLines(false);
        this.histogram.getAxisLeft().setDrawAxisLine(false);
        this.histogram.getAxisLeft().setDrawLabels(false);
        this.histogram.getAxisLeft().setSpaceTop(30.0f);
        this.histogram.getAxisLeft().setAxisMinimum(0.0f);
        this.histogram.getAxisRight().setEnabled(false);
        this.histogram.getXAxis().setDrawGridLines(false);
        this.histogram.getXAxis().setDrawAxisLine(false);
        this.histogram.getXAxis().setDrawLabels(false);
        this.histogram.setViewPortOffsets((float) this.viewPortLeftOffset, 0.0f, (float) this.viewPortRightOffset, 0.0f);
        LineChart lineChart = this.histogram;
        GraphInactiveFillRenderer graphInactiveFillRenderer2 = new GraphInactiveFillRenderer(lineChart, lineChart.getAnimator(), this.histogram.getViewPortHandler());
        this.graphInactiveFillRenderer = graphInactiveFillRenderer2;
        graphInactiveFillRenderer2.inactiveColor = this.inactiveGraphColor;
        this.histogram.setRenderer(this.graphInactiveFillRenderer);
        Drawable drawable = this.activeGraphFillDrawable;
        if (drawable == null) {
            drawable = AppCompatResources.getDrawable(getContext(), R.color.color_fill_cubic);
        }
        this.graphGradientDrawable = drawable;
        this.histogramTransitionAnimator.setInterpolator(ANIMATION_INTERPOLATOR);
        this.histogramTransitionAnimator.setDuration(300);
    }

    public void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        if (this.histogramTransitionAnimator.isRunning()) {
            this.histogramTransitionAnimator.cancel();
        }
    }

    public void setSeekbarEnabled(boolean z) {
        //  this.rangeSeekbar.setEnabled(z);
    }

    public final void updateGraph(List<Entry> list, boolean z) {
        if (list.isEmpty()) {
//            this.graphInactiveFillRenderer.setForceInactive(true);
//            this.histogram.invalidate();
//            this.noListingsDisabledDummySeekbar.setVisibility(0);
//            this.rangeSeekbar.setVisibility(8);
//            RangeSeekBar<Double> rangeSeekBar = this.rangeSeekbar;
//            rangeSeekBar.setSelectedMinValue(rangeSeekBar.absoluteMinValue);
//            rangeSeekBar.setSelectedMaxValue(rangeSeekBar.absoluteMaxValue);
            return;
        }
//        this.noListingsDisabledDummySeekbar.setVisibility(8);
//        this.rangeSeekbar.setVisibility(0);
        LineData lineData = this.histogram.getLineData();
        String str = "com.seatgeek.pricegraphLABEL_DATA";
        LineDataSet lineDataSet = (LineDataSet) lineData.getDataSetByLabel(str, false);
        if (lineDataSet == null) {
            lineDataSet = new LineDataSet(null, str);
            lineDataSet.setColor(-1);
            lineDataSet.setDrawFilled(true);
            lineDataSet.setFillDrawable(this.graphGradientDrawable);
            lineDataSet.setDrawIcons(false);
            lineDataSet.setHighlightEnabled(false);
            lineDataSet.setDrawHighlightIndicators(false);
            lineDataSet.setDrawCircles(false);
            lineDataSet.setDrawCircleHole(false);
            lineDataSet.setLineWidth(0.0f);
            lineDataSet.setHighlightLineWidth(0.0f);
            lineDataSet.setMode(LineDataSet.Mode.CUBIC_BEZIER);
            lineDataSet.setValueTextSize((float) this.graphValueTextSize);
            //lineDataSet.setValueFormatter(this.debugGraphFormatter);
            lineData.addDataSet(lineDataSet);
        }
        lineDataSet.setDrawValues(z);
        float[] fArr = new float[list.size()];
        for (int i = 0; i < list.size(); i++) {
            Entry entry = list.get(i);
            LineDataSet lineDataSet2 = this.previousDataSet;
            fArr[i] = (lineDataSet2 == null || lineDataSet2.getEntryCount() <= 0) ? 0.0f : this.previousDataSet.getEntryForXValue(entry.getX(), entry.getY()).getY();
        }
        this.histogramTransitionAnimator.removeAllListeners();
        ValueAnimator valueAnimator = this.histogramTransitionAnimator;
        float[] r4 = new ValueAnimator.AnimatorUpdateListener(lineDataSet, list, fArr, lineData) {
            private final /* synthetic */ LineDataSet f$1;
            private final /* synthetic */ List f$2;
            private final /* synthetic */ float[] f$3;
            private final /* synthetic */ LineData f$4;

            {
                this.f$1 = r2;
                this.f$2 = r3;
                this.f$3 = r4;
                this.f$4 = r5;
            }

            public final void onAnimationUpdate(ValueAnimator valueAnimator) {
                HistogramSeekbarView.this.lambda$updateGraph$2$HistogramSeekbarView(this.f$1, this.f$2, this.f$3, this.f$4, valueAnimator);
            }
        };
        valueAnimator.addUpdateListener(r4);
        this.histogramTransitionAnimator.addListener(new AnimatorListenerAdapter() {
            public final void onAnimationEnd(Animator animator) {
                HistogramSeekbarView.this.previousDataSet = lineDataSet;
            }

            public final void onAnimationCancel(Animator animator) {
                HistogramSeekbarView.this.previousDataSet = lineDataSet;
            }
        });
        if (this.histogramTransitionAnimator.isRunning()) {
            this.histogramTransitionAnimator.cancel();
        }
        this.histogramTransitionAnimator.start();
    }

    @SuppressLint("RestrictedApi")
    public void lambda$updateGraph$2$HistogramSeekbarView(LineDataSet lineDataSet, List list, float[] fArr, LineData lineData, ValueAnimator valueAnimator) {
        float animatedFraction = valueAnimator.getAnimatedFraction();
        lineDataSet.clear();
        int size = list.size();
        for (int i = 0; i < size; i++) {
            Entry entry = (Entry) list.get(i);
            lineDataSet.addEntry(new Entry(entry.getX(), AnimationUtils.lerp(fArr[i], entry.getY(), animatedFraction)));
        }
        lineData.notifyDataChanged();
        this.graphInactiveFillRenderer.setForceInactive(false);
        this.histogram.notifyDataSetChanged();
        this.histogram.invalidate();
    }

    public final void updateGraphInactiveFill(double d, double d2) {
        this.graphInactiveFillRenderer.activeMinXBound = (float) d;
        this.graphInactiveFillRenderer.activeMaxXBound = (float) d2;
        this.histogram.invalidate();
    }

    interface BinDeterminationDelegate {
        // Tuple2<Integer, BigDecimal> getOptimalBinInformation(List<BigDecimal> list, BigDecimal bigDecimal);
    }

    static class GraphInactiveFillRenderer extends LineChartRenderer {
        float activeMaxXBound;
        float activeMinXBound;
        int inactiveColor;
        private boolean forceInactive;

        GraphInactiveFillRenderer(LineDataProvider lineDataProvider, ChartAnimator chartAnimator, ViewPortHandler viewPortHandler) {
            super(lineDataProvider, chartAnimator, viewPortHandler);
        }

        /* access modifiers changed from: 0000 */
        public final void setForceInactive(boolean z) {
            this.forceInactive = z;
            if (z) {
                this.activeMinXBound = 0.0f;
                this.activeMaxXBound = 1.0f;
            }
        }

        public final void drawFilledPath(Canvas canvas, Path path, Drawable drawable) {
            ILineDataSet iLineDataSet = this.mChart.getLineData().getDataSetByLabel("com.seatgeek.pricegraphLABEL_DATA", false);
            int save = canvas.save();
            canvas.clipPath(path);
            canvas.drawColor(this.inactiveColor);
            if (!this.forceInactive) {
                Transformer transformer = this.mChart.getTransformer(iLineDataSet.getAxisDependency());
                drawable.setBounds((int) transformer.getPixelForValues(this.activeMinXBound, 0.0f).x, (int) this.mViewPortHandler.contentTop(), (int) transformer.getPixelForValues(this.activeMaxXBound, 0.0f).x, (int) this.mViewPortHandler.contentBottom());
                drawable.draw(canvas);
            }
            canvas.restoreToCount(save);
        }
    }

//    public void setListener(Listener listener2) {
//        this.listener = listener2;
//    }
}

File: cubic-chart/src/main/java/com/lyhv/cubic_chart/HistogramSeekbarView.kt
package com.lyhv.cubic_chart

import android.animation.TimeInterpolator
import android.animation.ValueAnimator
import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Path
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.View
import android.widget.LinearLayout
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.animation.addListener
import com.github.mikephil.charting.animation.ChartAnimator
import com.github.mikephil.charting.animation.Easing
import com.github.mikephil.charting.charts.LineChart
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import com.github.mikephil.charting.formatter.ValueFormatter
import com.github.mikephil.charting.interfaces.dataprovider.LineDataProvider
import com.github.mikephil.charting.renderer.LineChartRenderer
import com.github.mikephil.charting.utils.ViewPortHandler
import com.google.android.material.animation.AnimationUtils
import kotlinx.android.synthetic.main.histogram_seak_bar_view.view.slider
import java.text.NumberFormat

class HistogramSeekbarView : LinearLayout {
    private val histogramTransitionAnimator =
        ValueAnimator.ofFloat(0.0f, 1.0f).apply {
            interpolator = ANIMATION_INTERPOLATOR
        }
    private lateinit var histogram: LineChart
    var lastRealMax = 1.0
    var lastRealMin = 0.0

    /* access modifiers changed from: private */
    private var previousDataSet: LineDataSet? = null

    private var activeGraphFillDrawable: Drawable = AppCompatResources.getDrawable(
        context,
        R.drawable.histogram_fill_gradient
    )!!
    private lateinit var graphInactiveFillRenderer: GraphInactiveFillRenderer
    private val graphValueTextSize = 0
    private val inactiveGraphColor = Color.parseColor("#B5DCFA")
    private var lastSeekbarMaxValue = 1.0
    private var lastSeekbarMinValue = 0.0

    //  private RangeSeekBar noListingsDisabledDummySeekbar;
    private val viewPortLeftOffset = 0
    private val viewPortRightOffset = 0

    constructor(context: Context?) : super(context) {
        initializeHistogram()
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(
        context,
        attrs
    ) {
        initializeHistogram()
    }

    private fun initializeHistogram() {
        View.inflate(context, R.layout.histogram_seak_bar_view, this)
        histogram = findViewById(R.id.line_chart)
        histogram.data = LineData()
        histogram.setTouchEnabled(false)
        histogram.description = null
        histogram.setDrawGridBackground(false)
        histogram.setDrawBorders(false)
        histogram.setDrawMarkers(false)
        histogram.legend.isEnabled = false
        histogram.setNoDataText(null)
        histogram.setHardwareAccelerationEnabled(true)
        histogram.axisLeft.setDrawGridLines(false)
        histogram.axisLeft.setDrawAxisLine(false)
        histogram.axisLeft.setDrawLabels(false)
        histogram.axisLeft.spaceTop = 30.0f
        histogram.axisLeft.axisMinimum = 0.0f
        histogram.axisRight.isEnabled = false
        histogram.xAxis.setDrawGridLines(false)
        histogram.xAxis.setDrawAxisLine(false)
        histogram.xAxis.setDrawLabels(false)
        histogram.setViewPortOffsets(
            viewPortLeftOffset.toFloat(),
            0.0f,
            viewPortRightOffset.toFloat(),
            0.0f
        )
        graphInactiveFillRenderer = GraphInactiveFillRenderer(
            histogram,
            histogram.animator,
            histogram.viewPortHandler
        ).apply {
            inactiveColor = inactiveGraphColor
        }
        histogram.renderer = graphInactiveFillRenderer

        slider.addOnChangeListener { slider, value, fromUser ->
            updateProgressSeekBar(0.0, value.toDouble())
        }
    }

    public override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        if (histogramTransitionAnimator.isRunning) {
            histogramTransitionAnimator.cancel()
        }
    }

    fun updateGraph(
        entries: List<Entry>,
        z: Boolean
    ) {
        if (entries.isEmpty()) {
            graphInactiveFillRenderer.setForceInactive(true)
            histogram.invalidate()
            // TODO update seak bar here
            return
        }
        var lineDataSet =
            histogram.lineData.getDataSetByLabel(DATA_SET_LABEL, false) as? LineDataSet
        if (lineDataSet == null) {
            lineDataSet = LineDataSet(null, DATA_SET_LABEL)
            lineDataSet.color = -1
            lineDataSet.setDrawFilled(true)
            lineDataSet.fillDrawable = activeGraphFillDrawable
            lineDataSet.setDrawIcons(false)
            lineDataSet.isHighlightEnabled = false
            lineDataSet.setDrawHighlightIndicators(false)
            lineDataSet.setDrawCircles(false)
            lineDataSet.setDrawCircleHole(false)
            lineDataSet.lineWidth = 0.0f
            lineDataSet.highlightLineWidth = 0.0f
            lineDataSet.mode = LineDataSet.Mode.CUBIC_BEZIER
            lineDataSet.valueTextSize = graphValueTextSize.toFloat()
            lineDataSet.valueFormatter = CHART_VALUE_FORMATER
            lineDataSet.setDrawValues(z)
            histogram.data.addDataSet(lineDataSet)
        }
        if (this.previousDataSet == null) {
            this.previousDataSet = lineDataSet
        }
        val fArr = FloatArray(entries.size)

        entries.forEachIndexed { index, entry ->
            val lineDataSet2 = this.previousDataSet
            fArr[index] =
                if (lineDataSet2 == null || lineDataSet2.entryCount <= 0) 0.0f else previousDataSet?.getEntryForXValue(
                    entry.x,
                    entry.y
                )!!.y
        }
        val newFarr = entries.map {
            val dataSet = this.previousDataSet
            val invalidDataSet = dataSet == null || dataSet.entryCount <= 0
            if (invalidDataSet) {
                0.0f
            } else {
                dataSet!!.getEntryForXValue(
                    it.x,
                    it.y
                ).y
            }
        }
        logger<HistogramSeekbarView>().debug("${fArr[1]}")
        histogramTransitionAnimator.removeAllListeners()
        histogramTransitionAnimator.addListener(
            onEnd = {
                logger<HistogramSeekbarView>().debug("doOnEnd")
                previousDataSet = lineDataSet
            },
            onCancel = {
                logger<HistogramSeekbarView>().debug("doOnCancel")
                previousDataSet = lineDataSet
            }
        )
        histogramTransitionAnimator.addUpdateListener {
            updateChart(lineDataSet, entries, fArr, histogram.lineData, it)
        }

        if (this.histogramTransitionAnimator.isRunning) {
            this.histogramTransitionAnimator.cancel()
        }
        histogramTransitionAnimator.start()
    }

    @SuppressLint("RestrictedApi")
    private fun updateChart(
        lineDataSet: LineDataSet,
        list: List<Entry>,
        fArr: FloatArray,
        lineData: LineData,
        valueAnimator: ValueAnimator
    ) {
        val animatedFraction = valueAnimator.animatedFraction
        val newEntrys = arrayListOf<Entry>()
        newEntrys.addAll(list)
        lineDataSet.clear()
        val size = newEntrys.size
        for (i in 0 until size) {
            val entry = newEntrys[i]
            lineDataSet.addEntry(
                Entry(
                    entry.x,
                    AnimationUtils.lerp(
                        fArr[i],
                        entry.y,
                        animatedFraction
                    )
                )
            )
            logger<HistogramSeekbarView>().debug(
                "${Entry(
                    entry.x,
                    AnimationUtils.lerp(
                        fArr[i],
                        entry.y,
                        animatedFraction
                    )
                )}"
            )
        }
        lineData.notifyDataChanged()
        graphInactiveFillRenderer.setForceInactive(false)
        histogram.notifyDataSetChanged()
        histogram.invalidate()
    }

    fun updateGraphInactiveFill(d: Double, d2: Double) {
        graphInactiveFillRenderer.activeMinXBound = d.toFloat()
        graphInactiveFillRenderer.activeMaxXBound = d2.toFloat()
        histogram.invalidate()
    }

    fun updateProgressSeekBar(
        d: Double,
        d2: Double
    ) {
        updateGraphInactiveFill(d, d2)
        var doubleValue = d
        if (lastSeekbarMinValue == d) {
            doubleValue = lastRealMin
        }
        var doubleValue2 = d2
        if (lastSeekbarMaxValue == d2) {
            doubleValue2 = lastRealMax
        }
        lastSeekbarMinValue = d
        lastSeekbarMaxValue = d2
    }

    internal class GraphInactiveFillRenderer(
        lineDataProvider: LineDataProvider,
        chartAnimator: ChartAnimator,
        viewPortHandler: ViewPortHandler
    ) : LineChartRenderer(lineDataProvider, chartAnimator, viewPortHandler) {
        var activeMaxXBound = 0f
        var activeMinXBound = 0f
        var inactiveColor = 0
        private var forceInactive = false

        fun setForceInactive(z: Boolean) {
            forceInactive = z
            if (z) {
                activeMinXBound = 0.0f
                activeMaxXBound = 1.0f
            }
        }

        public override fun drawFilledPath(
            canvas: Canvas,
            path: Path,
            drawable: Drawable
        ) {
            val iLineDataSet = mChart.lineData.getDataSetByLabel(DATA_SET_LABEL, false)
            val save = canvas.save()
            canvas.clipPath(path)
            canvas.drawColor(inactiveColor)
            if (!forceInactive) {
                val transformer = mChart.getTransformer(iLineDataSet.axisDependency)
                drawable.setBounds(
                    transformer.getPixelForValues(
                        activeMinXBound,
                        0.0f
                    ).x.toInt(),
                    mViewPortHandler.contentTop().toInt(),
                    transformer.getPixelForValues(activeMaxXBound, 0.0f).x.toInt(),
                    mViewPortHandler.contentBottom().toInt()
                )
                drawable.draw(canvas)
            }
            canvas.restoreToCount(save)
        }
    }

    companion object {
        private const val DATA_SET_LABEL = "HistogramSeekbarView"
        private val ANIMATION_INTERPOLATOR: TimeInterpolator =
            Easing.EaseInOutSine
        private val INTEGER_FORMAT =
            NumberFormat.getIntegerInstance()
    }

    private object CHART_VALUE_FORMATER : ValueFormatter() {
        override fun getFormattedValue(
            value: Float,
            entry: Entry?,
            dataSetIndex: Int,
            viewPortHandler: ViewPortHandler?
        ): String {
            return INTEGER_FORMAT.format(value)
        }
    }

}

